src/java.base/share/classes/java/util/stream/ForEachOps.java

```diff
@@ -417,7 +426,9 @@ private static <S, T> void doCompute(ForEachOrderedTask<S, T> task) {
                 // Completion of the left child "happens-before" completion of
                 // the right child
                 rightChild.addToPendingCount(1);
-                task.completionMap.put(leftChild, rightChild);
+                // leftChild and rightChild were just created and not fork():ed
+                // yet so no need for a volatile write
+                NEXT.set(leftChild, rightChild);
```

Comment:
Make `next` a plain field and shuffle up the assignment (`leftChild.next = rightChild`) to occur immediately after construction of the right child task? (FWIW `addToPendingCount` operates on a volatile field of `CountedCompleter`).

Reply:
@PaulSandoz I'm usually a bit weary of piggybacking if it is not done on the same object, as future reorderings of the code might break that assumption. I wouldn't want to break anything silently so I made a rather conservative change. On which side should I err? :)

Reply:
Since as you have said the left and right nodes have yet to be "published" (outside of their little nest) I think it should be fine to move it above and next to the constructions.
(Also since `addToPendingCount` has volatile write semantics, via it's getAndSet, the plain write should not float below that call if that really matters)

Reply:
@PaulSandoz Sold! Making the changes :)

Reply:
@PaulSandoz Done!